File: hmvvm/src/main/java/com/see99/hmvvm/base/BaseFragment.java
package com.see99.hmvvm.base;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.fragment.app.Fragment;

import com.see99.hmvvm.MyApp;
import com.see99.hmvvm.dagger.component.ActivityComponent;
import com.see99.hmvvm.dagger.component.DaggerActivityComponent;
import com.see99.hmvvm.dagger.component.DaggerFragmentComponent;
import com.see99.hmvvm.dagger.component.FragmentComponent;

import javax.inject.Inject;

public abstract class BaseFragment<T extends ViewDataBinding, V extends BaseViewModel> extends Fragment {

    private BaseActivity mActivity;

    private T binding;

//    @Inject
    private V mViewModel;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof BaseActivity){
            this.mActivity = (BaseActivity) context;
        }
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        performInject(getFragmentComponent());
        super.onCreate(savedInstanceState);

    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater,getLayoutId(),container,false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        binding.setVariable(getBindingVariable(), mViewModel);
        binding.setLifecycleOwner(this);
        binding.executePendingBindings();
    }

    public abstract int getBindingVariable();

    @LayoutRes
    protected abstract int getLayoutId();

    //注入方法
    protected abstract void performInject(FragmentComponent fragmentComponent);

    private FragmentComponent getFragmentComponent() {
        return DaggerFragmentComponent.builder()
                .setAppComponent(((MyApp)(getContext().getApplicationContext())).appComponent)
                .fragment(this)
                .build();
    }
}

File: hmvvmkotlin/src/main/java/com/see99/hmvvmkotlin/base/BaseFragment.kt
package com.see99.hmvvmkotlin.base

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.fragment.app.Fragment
import com.see99.hmvvmkotlin.MyApp
import com.see99.hmvvmkotlin.dagger.component.ActivityComponent
import com.see99.hmvvmkotlin.dagger.component.DaggerActivityComponent
import com.see99.hmvvmkotlin.dagger.component.DaggerFragmentComponent
import com.see99.hmvvmkotlin.dagger.component.FragmentComponent
import javax.inject.Inject

abstract class BaseFragment<T : ViewDataBinding, V : BaseViewModel> : Fragment() {
    @Inject
    lateinit var viewModel: V
    lateinit var binding: T

    var activity: BaseActivity<*, *>? = null

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is BaseActivity<*,*>){
            activity = context
        }
        performInject(fragmentComponent())
    }

    private fun fragmentComponent(): FragmentComponent {

        return DaggerFragmentComponent.builder().fragment(this)
                .appComponent((this.context!!.applicationContext as MyApp).appComponent)
                .build()
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = DataBindingUtil.inflate(inflater,getLayoutId(),container,false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        performDataBinding()
        initBaseObserver()
        initView()
        initObserver()

    }
    abstract fun initView()

    abstract fun initObserver()

    private fun performDataBinding() {
        binding.setVariable(getVariableId(), viewModel)
        binding.lifecycleOwner = this
        binding.executePendingBindings()

    }

    abstract fun getVariableId(): Int

    @LayoutRes
    abstract fun getLayoutId(): Int

    private fun initBaseObserver() {
        viewModel.routers.observe(this) {
            if (it.bundle != null) {
                performRouter(it.routerId, it.bundle!!)
            } else {
                performRouter(it.routerId)
            }
        }

    }

    /**
     * 当前页可以跳转到多个页   可以根据routerId  进行判断   默认routerId 为 0
     */
    open fun performRouter(routerId: Int, bundle: Bundle) {
    }

    /**
     * 跳转页面不用传递参数
     */
    open fun performRouter(routerId: Int) {
        if (routerId == 0){
            performRouter()
        }
    }

    /**
     * routerid 默认为 0
     */
    open fun performRouter() {

    }

    /**
     * 注入方法
     */
    abstract fun performInject(fragmentComponent: FragmentComponent)

    private fun activityComponent(): ActivityComponent = DaggerActivityComponent.builder().build()
}